### ISSUE TITLE
port-forward closed on specific postgres command

### ISSUE THREAD

**Issue by Jeansen** [kubernetes/kubernetes#111825]
### What happened?

I've created a postgres cluster by the readme via helm and a simple manifest (see https://github.com/zalando).

I do `port-forward acid-minimal-cluster-0 8098:5432` and then `psql -h localhost -p 8098 -U postgres'. I paste the password from the kubernetes secret created by the operator.

While being connected via psql, I call `\c foo', the connection will then be closed with:

```
E0812 19:41:10.188388 2778901 portforward.go:406] an error occurred forwarding 8098 -> 5432: error forwarding port 5432 to pod 3902fddf18f9819a2114c71be1e745f7610b8695a8a3fa1c838525af3f1576e4, uid : port forward into network namespace "/var/run/netns/eeb57a40-33aa-4feb-a081-b3609a272a89": read tcp [::1]:42206->[::1]:5432: read: connection reset by peer
```

BUT, if I call the psql with a direct selection of a database, like so: `psql -h localhost -p 8098 -U postgres -d foo', I can do all sorts of calls, like select, create an so on.

### What did you expect to happen?

I'd expect no disconnection, no matter what command I run during a session with psql.

### How can we reproduce it (as minimally and precisely as possible)?

```console
git clone https://github.com/zalando/postgres-operator.git
helm repo add postgres-operator-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator0
helm install postgres-operator postgres-operator-charts/postgres-operator

#From within the git repo, above
kubectl create -f minimal-postgres-manifest.yaml

port-forward acid-minimal-cluster-0 8098:5432
#Install posgresql-client, e.g. sudo apt install postgresql-client
psql -h localhost -p 8098 -U postgres
#paste Password from kubernetes secret and then, while in qsql run
 \c foo
```

### Anything else we need to know?

Using `PGSSLMODE=disable psql -h localhost -p 8098 -U postgres` works so far, as there will be no connection drop, but uploads of larger files will hang.

Using the solution with a scat from here https://stackoverflow.com/questions/70762591/kubectl-port-forwarding-to-remote-postgres-port helped, but then data transfers are terribly slow!

### Kubernetes version

<details>

```console
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:30:46Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:23:26Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>
On premise
</details>

### OS version

<details>

```console
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>
crio 1.24.2
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
```console
Calico:
Client Version:    v3.23.3
Git commit:        3a3559be1
Cluster Version:   v3.23.3
Cluster Type:      typha,kdd,k8s,operator,bgp,kubeadm
```

PROMETHEUS_VERSION='14.12.0'
ROOK_VERSION='v1.9.3'
DASHBOARD_VERSION='5.4.1'
METRICS_VERSION='3.8.2'
METALLB_VERSION='0.12.1'
INGRESS_VERSION='4.1.1'

</details>

**Comment by brianpursley** [comment]
I'm not able to get your reproduction working for a couple of reasons:

First, it looks like maybe a typo in your reproduction commands:
```
helm repo add postgres-operator-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator0
```
should be this?
```
helm repo add postgres-operator-charts https://opensource.zalando.com/postgres-operator/charts/postgres-operator
```

Second, after I install the postgres-operator, it looks like it is running, I am able to port forward, but when I try to connect using `psql -p 8098 -h localhost -U postgres`, I get this error:
```
$ psql -p 8098 -h localhost -U postgres
psql: error: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

And my port forwarding session terminates like this:
```
$ kubectl port-forward postgres-operator-6776c88757-bnc5w 8098:5432
Forwarding from 127.0.0.1:8098 -> 5432
Forwarding from [::1]:8098 -> 5432
Handling connection for 8098
E0814 18:53:27.804820   30079 portforward.go:406] an error occurred forwarding 8098 -> 5432: error forwarding port 5432 to pod 39208642f47613e8cb33b6a17a22a3190c2297599e5a927933ca70227e729783, uid : failed to execute portforward in network namespace "/var/run/netns/cni-649d7884-a3a3-9a41-c72b-60c7ba695b8d": failed to connect to localhost:5432 inside namespace "39208642f47613e8cb33b6a17a22a3190c2297599e5a927933ca70227e729783", IPv4: dial tcp4 127.0.0.1:5432: connect: connection refused IPv6 dial tcp6: address localhost: no suitable address found 
E0814 18:53:27.805478   30079 portforward.go:234] lost connection to pod
```

Do you have any additional information you can provide to help reproduce this?

**Comment by brianpursley** [comment]
Incidentally, when I run postgres without helm, and do port forwarding it seems to work fine for me.

```
kubectl run postgres --image=postgres --env=POSTGRES_PASSWORD=hunter2
kubectl wait --for=condition=Ready pod/postgres
kubectl port-forward postgres 8098:5432
```

Then my psql run...
```
$ psql -p 8098 -h localhost -U postgres
psql (12.11 (Ubuntu 12.11-0ubuntu0.20.04.1), server 14.5 (Debian 14.5-1.pgdg110+1))
WARNING: psql major version 12, server major version 14.
         Some psql features might not work.
Type "help" for help.

postgres=# \c foo
FATAL:  database "foo" does not exist
Previous connection kept
postgres=# \c foo
FATAL:  database "foo" does not exist
Previous connection kept
postgres=# \list
                                 List of databases
   Name    |  Owner   | Encoding |  Collate   |   Ctype    |   Access privileges   
-----------+----------+----------+------------+------------+-----------------------
 postgres  | postgres | UTF8     | en_US.utf8 | en_US.utf8 | 
 template0 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
(3 rows)

postgres=# \q
~ $ 
```

My port forward session remains active and looks like this:
```
$ kubectl port-forward postgres 8098:5432
Forwarding from 127.0.0.1:8098 -> 5432
Forwarding from [::1]:8098 -> 5432
Handling connection for 8098
Handling connection for 8098
Handling connection for 8098
```

Can you try the above, and see if it works for you?

---

For refernece, here is my kubectl version output:
```
Client Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.3.254+e5a649d4f616bd", GitCommit:"e5a649d4f616bdbe6f500bbb2792ecb0a01c2b45", GitTreeState:"clean", BuildDate:"2022-07-28T23:44:02Z", GoVersion:"go1.18.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.3-40-g80c2a0fe3934cf.dev-1658410733-dirty", GitCommit:"80c2a0fe3934cfda6528118e54865eb13e011f48", GitTreeState:"dirty", BuildDate:"2022-07-21T13:40:54Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
```

and uname -a:
```
Linux cinlogic-xps13 5.15.0-46-generic #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

and my psql version:
```
psql (PostgreSQL) 12.11 (Ubuntu 12.11-0ubuntu0.20.04.1)
```

**Comment by panteparak** [comment]
@brianpursley @Jeansen From my usage. Noted that connection dropping issue is due to PostgreSQL is sending an RST Packet and somewhere in kubectl port forwarding it will drop the proxy once it see this packet.

This behavior is only seen while using an TLS version of postgresql. Hence the OP PGSSLMODE=disable env

**Comment by Jeansen** [comment]
@brianpursley thanks for your quick reply. I corrected my mistake. You are right, there was a typo. With regard to the connection: That's odd. When the cluster is up and running and you have the forward established, then it should work exactly as described and prompt you for a password.

If postgres itself works, then that's an interesting finding. On the other hand, it does not help much, because I cannot use it in my setup. Also, that's not the image used by the operator, which is: registry.opensource.zalan.do/acid/spilo-14:2.1-p6

**Comment by brianpursley** [comment]
Thank you @panteparak for reminding me about that. After you mentioned that, I recalled reading that elsewhere and I think you are right about the cause.

@Jeansen I was able to get your example to reproduce the problem using kind.  My bare metal test cluster must have been in some weird state.  Thanks for bearing with me.

What do people think about this PR?  It seems to fix the problem for me locally:
https://github.com/kubernetes/kubernetes/pull/111860

Basically, just because we get something on the error stream, doesn't mean it has to be fatal.  In some cases like connection reset by peer, it doesn't mean that the server process has been terminated, so it is fine (I think) to leave the port forwarding connection open and available to handle new requests.  The PR just handles that error as a special, non-fatal case.

**Comment by panteparak** [comment]
@brianpursley @Jeansen Just wanted to make sure If the introduced PR does not have a feature conflict with this PR that I believe introduced the RST Packet problem. #103526 

The original post that enlighten me of the RST Packet Issue is this. [kubectl Issue](https://github.com/kubernetes/kubectl/issues/1169#issuecomment-1165140134) 
Do read the entire thread. Some people were talking about having an additional cli argument to enabled/disabled such feature too.

**Comment by brianpursley** [comment]
@panteparak Good question, but I don't think it does.  The intention behind that PR was that usually when a message was sent on the error stream, the server connection was stopped and the port forwarding session remained open in an inoperable state, meaning you had to manually kill and restart the `kubectl port-forward` to get it working again.

I *think* "connection reset by peer" is a special case, where the port forwarding session should not be abandoned, as the server is still listening for new connections, just not handling that particular connection anymore. The pod is still running.

Maybe someone with a deeper background in networking could chime in, but I think it is probably wrong to assume that anything we get on the error stream is automatically a fatal error. Often times it may be, but in this case not.

**Comment by Jeansen** [comment]
@brianpursley I agree. For me the argumentation is sound. If we port-forward to a service but that service does whatever it wants regarding its connection, e.g. sending an HTTP 404, 500 or some sort of peer problem, then I too see the problem with the service, but not with the port forwarding. So, I would also vote for a robust port-forward that a user has to kill manually, if a user thinks the connections is in a bad shape but which stays alive otherwise.

**Comment by brianpursley** [comment]
> Do read the entire thread. Some people were talking about having an additional cli argument to enabled/disabled such feature too.

The flag would be OK, but it still has the original problem where if the pod restarts, your port forwarding session gets stuck in limbo and will never work again until you kill and rerun the command.

I'm open to the flag, but my thought (right now at least) is to handle "connection reset by peer" as a non fatal error, and if that solves it, then the flag would not be necessary.

I'm curious to see what the reviewer's have to say about the PR and this approach.

**Comment by aojea** [comment]
> I _think_ "connection reset by peer" is a special case, where the port forwarding session should not be abandoned, as the server is still listening for new connections, just not handling that particular connection anymore. The pod is still running.

that is only one scenario, you can receive a "connection reset by peer" when there is nothing listening on that port

**Comment by aojea** [comment]
> E0812 19:41:10.188388 2778901 portforward.go:406] an error occurred forwarding 8098 -> 5432: error forwarding port 5432 to pod 3902fddf18f9819a2114c71be1e745f7610b8695a8a3fa1c838525af3f1576e4, uid : port forward into network namespace "/var/run/netns/eeb57a40-33aa-4feb-a081-b3609a272a89": read tcp [::1]:42206->[::1]:5432: read: connection reset by peer


let me try to remember how this works how this works is:

client --- [listener] [portforward] ----- [apiserver] -------- [kubelet][CRI] ------- [containerd/crio] ----- [pod] --- process(psql here)
                     \ kubectl /


the error in the description is the stack trace , and IIUIC the reset by peer is the tunnel opened by crio/containerd inside the pod to postgresql (in this example)

We should check this retries that we are adding here carefully to avoid leaking connection, there were already some bugs related to this ....

**Comment by brianpursley** [comment]
```
psql --- [listener] [portforward] ----- [apiserver] -------- [kubelet][CRI] ------- [containerd/crio] ----- [pod] --- process(postgres here)
                     \ kubectl /
```

> We should check this retries that we are adding here carefully to avoid leaking connection, there were already some bugs related to this ....

I probably wasn't clear enough.  This isn't adding retry logic.  It is simply not terminating the listener (pf.streamConn) if the error was ECONNRESET.  The listener will remain running and able to service new psql commands, as opposed to terminating and requiring the user to restart port forwarding over and over again.

**Comment by aojea** [comment]
> probably wasn't clear enough. This isn't adding retry logic. It is simply not terminating the listener (pf.streamConn) if the error was ECONNRESET.

yeah, thanks for clarifying, I figured out later , sorry :smile:

**Comment by thockin** [comment]
Port-fwd is not aig-net, oddly

**Comment by mezhaka** [comment]
I cannot give a succinct repro setup at the moment, but I can try to do it upon request.

I am not doing any PostgreSQL, but connect to an internal service. Our CI test code
1) Creates a pod with the service.
2) Waits for that pod to go into a "Running" state.
3) Opens a tunnel.
4) Sends `HEAD /` requests to the service until it starts responding in order to figure out when it has finished booting (it takes about 15 seconds to boot and we probe twice per second).
5) Sends meaningful requests to the service.

Looking at Wireshark dumps, I confirm that while doing item 4) at some point an RST comes from the tunnel at which point the tunnel closes. If instead item 4) I just wait for some hard coded amount of time (I know that the service takes about 15 seconds to boot) then tunnel does not collapse.

**Comment by mezhaka** [comment]
I got a short repro based on your echo server.

1) do `kubectl run tcpecho --image=alpine --restart=Never -- /bin/sh -c "sleep 60 && apk add socat && socat -v tcp-listen:8080,fork EXEC:cat"`, note the `sleep 60 part`
2) start a tunnel `kubectl port-forward tcpecho 61630:8080` (I used 61630, to filter with Wireshark).
3) do a series of `nc -v localhost 61630`
4) observe that for the kubectl 1.22.6, the tunnel prints an error when I do 3), but the process does not die, so I can continue to do 3) until `tcpecho` starts and I get connected.
5) observe that for the kubectl 1.23.6, the tunnel prints an error and the process exists.

The "normal" mode of operation for us was 4), now this is failing after we recently updated to a newer kubectl.

I also do not see any RSTs in the Wireshark, just some [FIN, ACK].

A few questions for @brianpursley: 
1) Do you think the https://github.com/kubernetes/kubernetes/pull/111860 will enable our "normal" mode of operation (even no RSTs are on the wire)?
2) Is there an approximate estimate when will it be merged?

**Comment by brianpursley** [comment]
> I observe that kubectl 1.22.6 is not terminating when the RST as opposed to 1.23.6, however even though 1.22.6 is still running, I am not able to connect through it anyway.

Yeah, this was the reason for [this fix to make kubectl port-forward terminate on error](https://github.com/kubernetes/kubernetes/pull/103526). Otherwise, as you discovered, port forwarding appears to still be running, but is in some kind of zombie state where it cannot actually forward anymore.

> Do you think the https://github.com/kubernetes/kubernetes/pull/111860 will enable our "normal" mode of operation (even no RSTs are on the wire)?

I tried to reproduce using the tcpecho example you gave and on step 3 (`nc -v localhost 61630`) I don't get either an error or an exit, as you described in steps 4 and 5)

Is there anything you left out?  I'm just connecting to nc, typing some things, and ctrl+c to stop.

That being said, #111860 only handles "connection reset by peer" error, so if you're not seeing an RST packet, it would not have an effect.

**Comment by mezhaka** [comment]
@brianpursley First of all, thank you a lot for looking into this!

> Is there anything you left out? I'm just connecting to nc, typing some things, and ctrl+c to stop.
I have prepended `sleep 60`, so I have some time to create the port forward and then `nc`. The idea is to try to `nc` while the echo server is not yet running, so if you already have it running and port forward to it, you won't see what I see. Delete your tcpecho pod first.

I just tried it again. First I prepare my commands, without running them:
1) first shell tab:
```
nc -v localhost 61630
```
2) second shell tab:
```
kubectl port-forward tcpecho 61630:8080
```
3) third shell tab:
```
kubectl run tcpecho --image=alpine --restart=Never -- /bin/sh -c "sleep 60 && apk add socat && socat -v tcp-listen:8080,fork EXEC:cat"
```
Now I run 3), then 2), then 1).
In 1) I see:
```
$ nc -v localhost 61630
Connection to localhost port 61630 [tcp/*] succeeded!
```
In 2) I see:
```
$ kubectl port-forward tcpecho 61630:8080
Forwarding from 127.0.0.1:61630 -> 8080
Forwarding from [::1]:61630 -> 8080
Handling connection for 61630
E0823 20:24:30.226995   10692 portforward.go:406] an error occurred forwarding 61630 -> 8080: error forwarding port 8080 to pod 7ad31c9a5f2ccb941b0d6726487ab536ba73e578f1ac6dceab1acd468405d31e, uid : failed to execute portforward in network namespace "/var/run/netns/cni-4877b844-513e-d1f6-5428-2630e9fa38b1": failed to connect to localhost:8080 inside namespace "7ad31c9a5f2ccb941b0d6726487ab536ba73e578f1ac6dceab1acd468405d31e", IPv4: dial tcp4 127.0.0.1:8080: connect: connection refused IPv6 dial tcp6 [::1]:8080: connect: cannot assign requested address
E0823 20:24:30.227655   10692 portforward.go:234] lost connection to pod
```

**Comment by aojea** [comment]
There is nothing listening on that port inside the pod and it fails ... What's the expectation ?

**Comment by mezhaka** [comment]
@aojea As a user I would find it convenient if my `nc` command would fail with timeout or with some connection error, but not the tunnel. This would allow me to open a tunnel once and retry the command that goes through the tunnel (`nc` in my example), while right now both the tunnel and the command fail and I need to try to establish a tunnel again.

**Comment by aojea** [comment]
yeah, but shouldn't you have a probe to the port in the pod to indicate that is ready?
the port-forward will not try to connect to that pod

**Comment by EraYaN** [comment]
The pod is running in that example the (correct) process in the pod is not though. It it just to simulate any failed connection as an example. Postgres just sends RST packets by design, and that causes the same behavior but is harder to reproduce and much heavier to setup.

**Comment by EraYaN** [comment]
This is not fixed, at best it's worked around by disabling `sslmode`.

/remove-lifecycle stale

**Comment by aojea** [comment]
> Postgres just sends RST packets by design, and that causes the same behavior but is harder to reproduce and much heavier to setup.

The RST is not a postgres thing , IIUIC it is because portforward tries to connect to a socket that is not already open, and the system sends a RST

**Comment by aojea** [comment]
ok, If I understand this correctly, the expected behavior for this feature is as an ssh -L9090:localhost:8080, so I keep the tunnel open and the client deals with the network errors of nothing listening on the other side, is that correct?

/cc @brianpursley

**Comment by brianpursley** [comment]
> The RST is not a postgres thing

You are correct, this is not a problem in Postgres itself, and RST is not a Posgres thing.  It's just that this problem is very noticeable when using Posgres w/ SSL because the lost connection makes it unusable. Presumably other applications have the same problem, but maybe not as severe.

> IIUIC it is because portforward tries to connect to a socket that is not already open, and the system sends a RST

The behavior I *think* I've seen is the connection is opened successfully, some data is sent and received successfully, but then at some point data is sent and "connection reset by peer" comes back on the error stream.

In this PR comment, I provided some additional context about the problem and where I *think* it originates from: 
 https://github.com/kubernetes/kubernetes/pull/111860#issuecomment-1245862906

> ok, If I understand this correctly, the expected behavior for this feature is as an ssh -L9090:localhost:8080, so I keep the tunnel open and the client deals with the network errors of nothing listening on the other side, is that correct?

I'm not sure if this is demonstrable using ssh or not, but I think what you have said is correct.  We want the tunnel to remain open and only the handled connection to be closed in this case, not the overall tunnel.
